common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,893 @@
+      this.appId = appId;
+      this.shuffleId = shuffleId;
+      this.reduceId = reduceId;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      AppShufflePartitionId that = (AppShufflePartitionId) o;
+      return shuffleId == that.shuffleId && reduceId == that.reduceId
+        && Objects.equal(appId, that.appId);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(appId, shuffleId, reduceId);
+    }
+
+    @Override
+    public String toString() {
+      return Objects.toStringHelper(this)
+        .add("appId", appId)
+        .add("shuffleId", shuffleId)
+        .add("reduceId", reduceId)
+        .toString();
+    }
+
+    String generateFileName() {
+      return String.format("mergedShuffle_%s_%d_%d", appId, shuffleId, reduceId);
+    }
+
+    boolean compareAppShuffleId(String appId, int shuffleId) {
+      return Objects.equal(this.appId, appId) && this.shuffleId == shuffleId;
+    }
+  }
+
+  /** Metadata tracked for an actively merged shuffle partition */
+  public static class AppShufflePartitionInfo {
+
+    private final AppShufflePartitionId partitionId;
+    // The merged shuffle data file
+    final File dataFile;
+    public final FileChannel channel;
+    // Location offset of the last successfully merged block for this shuffle partition
+    private long position;
+    // Indicating whether failure was encountered when merging the previous block
+    private boolean encounteredFailure;
+    // Track the map Id whose block is being merged for this shuffle partition
+    private int currentMapId;
+    // Bitmap tracking which mapper's blocks have been merged for this shuffle partition
+    private RoaringBitmap mapTracker;
+    // The index file for a particular merged shuffle contains the chunk offsets.
+    private final FileChannel indexChannel;
+    // The meta file for a particular merged shuffle contains all the map ids that belong to every
+    // chunk. The entry per chunk is a serialized bitmap.
+    private final FileChannel metaChannel;
+    private final DataOutputStream indexWriteStream;
+    // The offset for the last chunk tracked in the index file for this shuffle partition
+    private long lastChunkOffset;
+    private int lastMergedMapId = -1;
+
+    // Bitmap tracking which mapper's blocks are in the current shuffle chunk
+    private RoaringBitmap chunkTracker;
+    ByteBuf trackerBuf = null;
+
+    AppShufflePartitionInfo(
+        AppShufflePartitionId partitionId,
+        File dataFile,
+        File indexFile,
+        File metaFile) throws IOException {
+      this.partitionId = Preconditions.checkNotNull(partitionId, "partition id");
+      dataFile.createNewFile();
+      this.dataFile = dataFile;
+      this.channel = new FileOutputStream(dataFile, true).getChannel();
```

Comment:
This is a miss in the earlier implementation.
The `append = true` is not necessary since we will not append into a merged shuffle file that's already finalized.

Reply:
If append is false (default), we are overwriting the file : why do we need `createNewFile` ?
FileOutputStream will create the file if missing.

Reply:
@mridulm You are right. We don't to use `createNewFile`. I am going to fix this.

Reply:
There is one scenario where we do need to open the file with append mode.
When we add support for levelDB state store to support NM restart, we will need to open these files again with append mode.

Reply:
I have made these changes.
@Victsm We can add the ability to write in append mode when NM restart is handled. `createNewFile` in that case will also not be needed.

Reply:
`createNewFile` could be used to atomically open new file : if we are checking return value.
Otherwise, as @otterc mentioned, `FileOutputStream` handles the various cases we need (either with `append` true or false)